sphinx/writers/html.py

```diff
@@ -863,6 +866,83 @@ def visit_math(self, node, math_env=''):
+            self.collected_footnotes.append(node)
+            raise nodes.SkipChildren
+        else:
+            BaseTranslator.visit_footnote(self, node)
+
+    def _extend_footnote_reference(self, node):
+        for _id in node['ids']:
+            for ref in self.footnote_refs.pop(_id):
+                self.body[ref] = self.body[ref].replace(
+                    '>', u(' title="{}">').format(
+                        node[1].astext().replace('\n', ' ')))
+
+    # overwritten
+    def depart_footnote(self, node):
+        if 'bottom' in self.builder.config.html_footnotes_options:
+            pass
+        else:
+            BaseTranslator.depart_footnote(self, node)
+
+    # overwritten
+    def depart_document(self, node):
+        if self.collected_footnotes:
+            self.build_footnotes()
+        BaseTranslator.depart_document(self, node)
+
+    def build_footnotes(self):
+        footnotes = self.document.copy()
+        footnotes.extend(self.collected_footnotes)
+        visitor = _BottomFootnotesTranslator(self.builder, footnotes)
+        footnotes.walkabout(visitor)
+        self.body.extend(visitor.body)
+
+    def visit_footnote_reference(self, node):
+        BaseTranslator.visit_footnote_reference(self, node)
+        if 'tooltip' in self.builder.config.html_footnotes_options:
+            self.footnote_refs[node['refid']].append(len(self.body) - 1)
+
     def unknown_visit(self, node):
         # type: (nodes.Node) -> None
         raise NotImplementedError('Unknown node: ' + node.__class__.__name__)
+
+
+class _BottomFootnotesTranslator(HTMLTranslator):
+    def visit_document(self, node):
+        self.body.append(self.starttag(node, 'div', CLASS='footnotes'))
+        if 'rubric' in self.builder.config.html_footnotes_options:
+            rubric_node = nodes.rubric(text=_('Footnotes'))
+            node.insert(0, rubric_node)
+        elif 'title' in self.builder.config.html_footnotes_options:
+            self.body.append('<h2>' + _('Footnotes') + '</h2>')
+            self._start_footnotes_table()
+        else:
+            self._start_footnotes_table()
+
+    def _start_footnotes_table(self):
+        self.body.extend(('<table class="docutils footnote" '
+                          'frame="void" rules="none">',
+                          '<colgroup><col class="label" /><col /></colgroup>',
+                          '<tbody valign="top">'))
+
+    def depart_document(self, node):
+        self.body.extend(('</tbody>', '</table>', '</div>'))
+
+    def visit_footnote(self, node):
+        self.body.append('<tr ' + ' '.join('id="{}"'.format(x)
+                                           for x in node['ids']) + '>')
+        self.footnote_backrefs(node)
+
+    def depart_footnote(self, node):
+        self.body.extend(('</td>', '</tr>'))
+
+    def depart_rubric(self, node):
+        HTMLTranslator.depart_rubric(self, node)
+        self._start_footnotes_table()
```

Comment:
Does this translator handle rubric which users written?

Reply:
the user should either define `rubric` in the new option or put a `rubric` directive at the end of a document wheer s_he knows that there are footnotes.